torch/csrc/distributed/rpc/rpc_agent.cpp

```diff
@@ -106,8 +106,20 @@ void RpcAgent::retryExpiredRpcs() {
       auto& earliestRpc = *it;
       // Making a copy of the message so it can be retried in the future.
       Message msgCopy = earliestRpc->message_;
-      auto fm = send(earliestRpc->to_, std::move(msgCopy));
-      futures.emplace_back(fm, earliestRpc);
+      std::shared_ptr<FutureMessage> fm;
+
+      // send() will throw an exception if an RPC is retried while the agent is
+      // shutdown. We must catch this exception and mark the original future
+      // with an error, since this RPC never succeeded and can no longer be
+      // retried.
+      try {
+        fm = send(earliestRpc->to_, std::move(msgCopy));
+        futures.emplace_back(fm, earliestRpc);
+      } catch (std::exception& e) {
+        std::string errorMsg =
+            c10::str("send() failed while retrying RPC with error: ", e.what());
+        earliestRpc->originalFuture_->setError(errorMsg);
```

Comment:
calling setError() with the lock held might lead to deadlocks.
(not sure if that's related to the observed issue)

Reply:
Thanks, just realized this as well. Not related to the test failure, but I'll push a fix